### ISSUE TITLE
Dealing with datasets when using Horovod on k8s cluster

### ISSUE THREAD

**Issue by raz-bn** [horovod/horovod#1715]
Hey! I would like to provide Horovod infrastructure on my organization GPU k8s cluster.
The only issue I struggle with is to find a solution on how to deal with large datasets.
I looked over some other issues and saw you are recommending to split the datasets into smaller datasets #249 #166 and then read them at random order.

However, I would like that my researchers won't have to deal with it, so I came up with the idea to create a small framework lets call it a data loader. when given a full dataset(path), each Horovod process will get a random split of the dataset to train with.

the bottom line is, I want to find a way to make the work with datasets as transparent as possible (the more the better) when using Horovod

### Q1
Is this idea even possible and will it be supported by the DL frameworks that Horovod supports,
or is there any other solution to my problem out there? 

### Q2
Is it better to copy the split datasets to the docker image when building it and then make each process read different parts of the dataset in random order or store the split datasets lets say in s3 and read them from the python script at random order (e.g. with boto3)? (or any other way you can suggest to me)

**Comment by tgaddair** [comment]
Hey @raz-bn, to answer your first question, we commonly recommend using the [petastorm](https://github.com/uber/petastorm) framework for reading data to automatically handle sharding, partitioning, shuffling, etc.  I would suggest taking a look at that first and let us know if you have any thoughts on how it might be improved to fit your use case.

For your second question, you generally do not want to pack subsets of the data into the image.  You would typically want to use the same image for all workers, and datasets can be quite big, so you're better off using something like S3 to store the data.

The way petastorm handles this (I recommend reading their blog post [here](https://eng.uber.com/petastorm/)) is to store small groups of examples into Parquet row groups, then randomly shuffle and partition the row groups among the workers.  You wouldn't want to pre-split the data because (1) your number of workers (thus, splits) might change, and (2) in order to improve convergence, you'll want to shuffle examples between workers after each epoch.

**Comment by raz-bn** [comment]
@tgaddair Thanks for the replay! 
So just let me know if I get you right. 
I'm going to generate the Petastorm dataset (parquet file), store in S3 and read it with the Petastrom API (sharding and shuffling) and that should do to trick? 

Can you provide details about the networking consequences of such an approach?

**Comment by tgaddair** [comment]
Hi @raz-bn, that's the general idea, yes.  We have an example [here](https://github.com/horovod/horovod/blob/master/examples/keras_spark_rossmann_run.py#L453) that shows you what the API would look like for Horovod + TensorFlow.

@selitvin, who owns the Petastorm project, could give you more details about the network implications.  Because Petastorm makes good use of caching and prefetching, you should not be bottlenecked by I/O in most cases.

**Comment by raz-bn** [comment]
Thanks! @tgaddair 
is there any slack channel for Horvod or Petastorm?

**Comment by tgaddair** [comment]
No problem @raz-bn.  I'm not sure about Petastorm, but we have a slack channel for Horovod at horovod-ai.slack.com.

**Comment by raz-bn** [comment]
@tgaddair seems like I need an invitation to get into your slack channel

**Comment by tgaddair** [comment]
@raz-bn do you have an email I can use to send you an invite?

**Comment by raz-bn** [comment]
@tgaddair Yes, Im going to send you an email about it (for privacy reasons)
